api/types/sandbox.proto

```diff
@@ -0,0 +1,51 @@
+/*
+	Copyright The containerd Authors.
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+syntax = "proto3";
+
+package containerd.types;
+
+import weak "gogoproto/gogo.proto";
+import "google/protobuf/any.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/containerd/containerd/api/types;types";
+
+// Sandbox represents a sandbox metadata object that keeps all info required by controller to
+// work with a particular instance. containerd keeps this in metadata store and sends it to a controller with
+// each request, so they can reflect sandbox state changes (by updating the corresponding fields).
+message Sandbox {
+	// SandboxID is a unique instance identifier within namespace
+	string sandbox_id = 1;
+	message Runtime {
+		// Name is the name of the runtime.
+		string name = 1;
+		// Options specify additional runtime initialization options.
+		google.protobuf.Any options = 2;
```

Comment:
These fields utilize existing channels to pass configuration that available today in containerd. I didn't introduce anything new, so clients can operate with familiar terms.
- Runtime options is something that you can pass to `StartShim` (though it wasn't used in the v2 runtime before, it is now exposed and can be, see details here https://github.com/containerd/containerd/pull/5368)
- Spec behaves same way as OCI spec, is is written to `config.json` in the bundle directory (see https://github.com/containerd/containerd/blob/master/runtime/v2/bundle.go#L103)
- Extensions are not for runtime, but rather for clients to store additional metadata (same as `Container.Extensions` that currently used by CRI)

Because it is expected to always have 1 shim = 1 sandbox = 1 VM instance, technically you can use both runtime options and spec. I'd love "spec" part to be standardized similarly to OCI spec (or even used directly, because it already includes bunch of fields relevant for VMs - https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L29). While runtime options will likely remain arbitrary and be runtime shim specific.